internal/writeaheadlog/wal.go

```diff
@@ -0,0 +1,295 @@
+	if err != nil {
+		return fmt.Errorf("saving value to WAL: %w", err)
+	}
+	err = wal.active.file.Sync()
+	if err != nil {
+		return fmt.Errorf("sycning the file: %w", err)
+	}
+
+	wal.active.maxEpoch = max(wal.active.maxEpoch, PT(&value).WALEpoch())
+	wal.active.content = append(wal.active.content, value)
+
+	return nil
+}
+
+// Purge removes files containing entries only order than keepEpoch
+// It is not guarnateed to remove all entries older than keepEpoch as it will keep the current
+// file.
+func (wal *WriteAheadLog[T, PT]) Purge(keepEpoch uint64) error {
+	wal.lk.Lock()
+	defer wal.lk.Unlock()
+
+	var keptLogFiles []logContent[T, PT]
+	var err error
+	for _, c := range wal.logFiles {
+		if c.maxEpoch < keepEpoch {
+			err1 := os.Remove(filepath.Join(wal.path, c.logname))
+			err = multierr.Append(err, fmt.Errorf("removing WAL file %q: %w", c.logname, err1))
+
+			continue
+		}
+		keptLogFiles = append(keptLogFiles, c)
+	}
+	wal.logFiles = keptLogFiles
+	return nil
+}
+
+const roatateAt = 1 << 20 // 1MiB
+
+func (wal *WriteAheadLog[T, PT]) maybeRotate() error {
+	if wal.active.file == nil {
+		return wal.rotate()
+	}
+	stats, err := wal.active.file.Stat()
+	if err != nil {
+		return fmt.Errorf("collecting stats for the file: %w", err)
+	}
+	if stats.Size() > roatateAt {
+		return wal.rotate()
+	}
+	return nil
+}
+
+const walExtension = ".wal.cbor"
+
+// Finalize closes the existing file
+// the WAL is safe to discard or can be used still
+func (wal *WriteAheadLog[T, PT]) Finalize() error {
+	wal.lk.Lock()
+	defer wal.lk.Unlock()
+	return wal.finalize()
+}
+
+// rotate a log file
+// it finalizes an active file if it is open
+// and opens a new one
+func (wal *WriteAheadLog[T, PT]) rotate() error {
+	if wal.active.file != nil {
+		if err := wal.finalize(); err != nil {
+			return fmt.Errorf("finalizing log file: %w", err)
+		}
+	}
+
+	wal.active.logname = time.Now().UTC().Format(time.RFC3339Nano) + walExtension
+	var err error
+	wal.active.file, err = os.OpenFile(
+		filepath.Join(wal.path, wal.active.logname),
+		os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0666)
+
+	if err != nil {
+		wal.active.file = nil
+		return fmt.Errorf("opening new log file %q: %w", wal.active.logname, err)
+	}
+	wal.active.cborWriter = cbg.NewCborWriter(wal.active.file)
+	return nil
+}
+
+// finalize the existing log
+func (wal *WriteAheadLog[T, PT]) finalize() error {
+	if wal.active.file == nil {
+		return nil
+	}
+	err := wal.active.file.Sync()
+	if err != nil {
+		return fmt.Errorf("syncing content to disk: %w", err)
+	}
+	err = wal.active.file.Close()
+	if err != nil {
+		return fmt.Errorf("closing the file: %w", err)
+	}
+	wal.active.file = nil
+	wal.active.cborWriter = nil
+
+	wal.active.logContent.complete = true
+	wal.logFiles = append(wal.logFiles, wal.active.logContent)
+	wal.active.logContent = logContent[T, PT]{}
+	return nil
+}
+
+func (wal *WriteAheadLog[T, PT]) readInAll() error {
+	_, err := os.Stat(wal.path)
+	if errors.Is(err, os.ErrNotExist) {
```

Comment:
nit: just directly call `os.ReadDir`? If the directory does not exist, there's nothing to read in anyways.